### PR TITLE
[3.0][Translator] changed the visibility of the locale from protected to private.

### PR DESCRIPTION
**Pull request by aitboudad**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | yes |
| Deprecations? | yes |
| Fixed tickets | ~ |
| Tests pass? | yes |
| License | MIT |

### CODE REVIEW

**Comment by mpdude** [comment]
Thats something users will have to find in the `3.0` upgrade file, so no deprecation warning in earlier `2.x` versions. Is that expected? (No idea how one could warn users, though...)

**Comment by aitboudad** [comment]
@mpdude Not sure if there is a way to deprecate protected property the only thing came to my mind is to add a note in `UPGRADE-2.8.md`.

**Comment by iltar** [comment]
There's a way but you won't like it. Implement `__get()` which returns only the property required but also throws a deprecation notice.

**Comment by aitboudad** [comment]
@iltar no it won't work as expected, the `locale` will be considered as public property.

**Comment by mpdude** [comment]
The question is what the "Symfony standard way" of doing this looks like.

**Comment by hhamon** [comment]
Does this BC break really add a value to the codebase? I mean, is it really worth it to take the risk of breaking applications / libraries relying on this `$locale` property?

**Comment by aitboudad** [comment]
@hhamon making `locale` private can reduce the risk to use an invalid locale see [setLocale](https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/Translation/Translator.php#L153) and  also I would prefer the public methods instead (setLocale, getLocale).

**Comment by aitboudad** [comment]
ping @symfony/deciders

**Comment by stof** [comment]
and given that we added validations of invalid locales to avoid a security issue, forcing to go through the setter to update it seems logical to me.

:+1:

File: src/Symfony/Bundle/FrameworkBundle/Tests/Translation/TranslatorTest.php [hunk 1]
```diff
@@ -296,14 +296,3 @@ private function createTranslator($loader, $options, $translatorClass = '\Symfon
...
 }
-
-class TranslatorWithInvalidLocale extends Translator
-{
-    /**
-     * {@inheritdoc}
-     */
-    public function setLocale($locale)
-    {
-        $this->locale = $locale;
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Translation/TranslatorTest.php]: why remove this tests? a custom translator can still return an invalid locale if it overwrites `getLocale` instead (or getLocale and setLocale).

File: src/Symfony/Bundle/FrameworkBundle/Tests/Translation/TranslatorTest.php [hunk 2]
```diff
@@ -94,14 +94,14 @@ public function testTransWithCaching()
...
 
+    /**
+     * @expectedException \InvalidArgumentException
+     */
     public function testTransWithCachingWithInvalidLocale()
     {
         $loader = $this->getMock('Symfony\Component\Translation\Loader\LoaderInterface');
-        $translator = $this->getTranslator($loader, array('cache_dir' => $this->tmpDir), 'loader', '\Symfony\Bundle\FrameworkBundle\Tests\Translation\TranslatorWithInvalidLocale');
+        $translator = $this->getTranslator($loader, array('cache_dir' => $this->tmpDir));
         $translator->setLocale('invalid locale');
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Translation/TranslatorTest.php]: this tests that setLocale doesn't accept invalid locales. but `getCatalogue` and all the other public methods are not covered.

**Comment by Tobion** [comment]
I'm :+1: given that it usually doesn't make sense to make properties protected when there are public getters/setters also defined. There is no point it creating 2 extension points.

File: src/Symfony/Bundle/FrameworkBundle/Tests/Translation/TranslatorTest.php (hunk 1 above)
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Translation/TranslatorTest.php]: reverted

File: src/Symfony/Bundle/FrameworkBundle/Tests/Translation/TranslatorTest.php (hunk 2 above)
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Translation/TranslatorTest.php]: I thought about moving this test into Translator component but it should be in a separate PR.

**Comment by xabbuh** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @aitboudad.
